sway-lsp/src/core/traverse_parse_tree.rs

```diff
@@ -0,0 +1,420 @@
+#![allow(dead_code)]
+
+use crate::{
+    core::token::{AstToken, TokenMap, TokenType},
+    utils::token::{desugared_op, to_ident_key},
+};
+use sway_core::{
+    constants::{MATCH_RETURN_VAR_NAME_PREFIX, TUPLE_NAME_PREFIX},
+    parse_tree::MethodName,
+    AstNode, AstNodeContent, Declaration, Expression, FunctionDeclaration, ReassignmentTarget,
+    TypeInfo, WhileLoop,
+};
+use sway_types::Ident;
+
+pub fn traverse_node(node: &AstNode, tokens: &mut TokenMap) {
+    match &node.content {
+        AstNodeContent::Declaration(declaration) => handle_declaration(declaration, tokens),
+        AstNodeContent::Expression(expression) => handle_expression(expression, tokens),
+        AstNodeContent::ImplicitReturnExpression(expression) => {
+            handle_expression(expression, tokens)
+        }
+        AstNodeContent::ReturnStatement(return_statement) => {
+            handle_expression(&return_statement.expr, tokens)
+        }
+        AstNodeContent::WhileLoop(while_loop) => handle_while_loop(while_loop, tokens),
+        // TODO
+        // handle other content types
+        _ => {}
+    };
+}
+
+fn handle_function_declation(func: &FunctionDeclaration, tokens: &mut TokenMap) {
+    tokens.insert(
+        to_ident_key(&func.name),
+        TokenType::from_parsed(AstToken::FunctionDeclaration(func.clone())),
+    );
+    for node in &func.body.contents {
+        traverse_node(node, tokens);
+    }
+    for parameter in &func.parameters {
+        tokens.insert(
+            to_ident_key(&parameter.name),
+            TokenType::from_parsed(AstToken::FunctionParameter(parameter.clone())),
+        );
+    }
+
+    handle_custom_type(&func.return_type, tokens);
+}
+
+fn handle_custom_type(_type_info: &TypeInfo, _tokens: &mut TokenMap) {
+    // TODO: Not obvious how to handle this now with the new types
```

Comment:
```suggestion
    todo!("Not obvious how to handle this now with the new types");
```

If we don't want to panic here, we should maybe open an issue for this and mention it under this `TODO` and move the commented out code to the issue?

Reply:
Good idea, I've opened #2274 and moved the code over in there for the time being.